packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx

```diff
@@ -30,25 +31,54 @@ import { ComboBoxInput } from './ComboBoxInput';
 import { ComboBoxInputContainer } from './ComboBoxInputContainer';
 import { ComboBoxMenu } from './ComboBoxMenu';
 import { ComboBoxMenuItem } from './ComboBoxMenuItem';
+import { ComboBoxMultipleContainer } from './ComboBoxMultipleContainer';
 
-export type ComboBoxProps<T> = {
+const isMultipleEnabled = <T,>(
+  variable: (T | undefined) | (T[] | undefined),
+  multiple: boolean,
+): variable is T[] | undefined => {
+  return multiple;
+};
+
+// Type guard for multiple did not work
+const isSingleOnChange = <T,>(
+  onChange:
+    | (((value: T) => void) | undefined)
+    | (((value: T, event: SelectValueChangeEvent) => void) | undefined),
+  multiple: boolean,
+): onChange is ((value: T) => void) | undefined => {
+  return !multiple;
+};
+
+type ComboboxMultipleTypeProps<T> = {
   /**
-   * Array of options that populate the dropdown list.
+   * Enable multiple selection.
+   * @default false
    */
-  options: SelectOption<T>[];
+  multiple?: boolean;
   /**
    * Controlled value of ComboBox.
+   * If `multiple` is enabled, it is an array of values.
+   */
+  value?: T | T[];
+  /**
+   * Callback fired when selected value changes.
    */
-  value?: T;
+  onChange?:
+    | ((value: T) => void)
```

Comment:
What if get rid of this overload and just make `event` optional?

e.g. `onChange?: (value: T, event?: SelectValueChangeEvent) => void`


That is effectively what the type is currently anyway (for the user), so simplifying this would just make it easier for us to use inside the component and we could get rid of the `isSingleOnChange` type guard.

Reply:
I added both type guards back. It nice to have some types enforcements in code.